File: src/main/go/src/euler/algorithm/long.go
// COPYRIGHT (C) 2014 barreiro. All Rights Reserved.
// GoLang helper for Project Euler problems
package algorithm

import "math"

// Default value used as base for the numeric system. Used in methods that make array-based calculations. Default to the decimal system.
const DefaultRadix = 10

// Table for fast lookup of powers of 10
var pow10 = []int{
	1,
	10,
	100,
	1000,
	10000,
	100000,
	1000000,
	10000000,
	100000000,
	1000000000,
	10000000000,
	100000000000,
	1000000000000,
	10000000000000,
	100000000000000,
	1000000000000000,
	10000000000000000,
	100000000000000000,
	1000000000000000000,
}

// calculates an approximate of the square root
func IntSqrt(value int) int {
	result, one := 0, 1<<30

	// "one" starts at the highest power of four <= than the argument
	for one > value {
		one >>= 2
	}

	approx := value
	for ; one != 0; one, result = one>>2, result>>1 {
		if approx >= result+one {
			approx -= result + one
			result = result + (one << 1)
		}
	}

	// Rounding to nearest integer
	if approx > result {
		return result + 1
	} else {
		return result
	}
}

// the sum of all the numbers up to value
func ArithmeticSum(value int) int {
	return value * (value + 1) / 2
}

// Simple method to calculate the factorial of small values. No checks are performed. Use with caution.
func Factorial(l int) int {
	if l <= 2 {
		return l
	}
	return l * Factorial(l-1)
}

func Pow(base int, exp int) int {
	if base == 0 && exp == 0 {
		return 1
	}
	if base == 0 {
		return 0
	}
	if base == 1 {
		return base
	}
	if base == 2 {
		return 1 << uint(exp)
	}
	if base == 10 {
		return Pow10(exp)
	}

	if exp == 0 {
		return 1
	}
	if exp == 1 {
		return base
	}
	if exp == 2 {
		return base * base
	}
	return squaring(base, exp)
}

func squaring(base int, exp int) int {
	result := 1
	for sqrBase, sqrExp := base, exp; sqrExp != 0; sqrExp, sqrBase = sqrExp/2, sqrBase*sqrBase {
		if sqrExp%2 != 0 {
			result *= sqrBase
		}
	}
	return result
}

func Square(base int) int {
	return base * base
}

// Convenience method to calculate the power when in base 10.
func Pow10(exp int) int {
	if exp < len(pow10) {
		return pow10[exp]
	} else {
		return -2
	}
}

// --- //

func IsPalindrome(value int) bool {
	return isPalindromeDigits(ToDigits(value))
}

func isPalindromeDigits(digits []int) bool {
	for l := 0; l*2 < len(digits); l++ {
		if digits[l] != digits[len(digits)-l-1] {
			return false
		}
	}
	return true
}

// --- //

func ToDigits(value int) []int {
	return toDigitsRadix(value, DefaultRadix)
}

func toDigitsRadix(value int, radix int) []int {
	digits := make([]int, 0, 32/IntSqrt(radix))
	for ; value >= radix; value /= radix {
		digits = append(digits, value%radix)
	}
	return append(digits, value)
}

// --- //

func Min(a int, b int) int {
	if a <= b {
		return a
	} else {
		return b
	}
}

func Max(a int, b int) int {
	if a >= b {
		return a
	} else {
		return b
	}
}

func MaxArray(a ...int) int {
	max := math.MinInt64
	for _, i := range a {
		if i > max {
			max = i
		}
	}
	return max
}

func MaxKey(a map[int]int) int {
	max := math.MinInt64
	for i := range a {
		if i > max {
			max = i
		}
	}
	return max
}

func ArraySum(a []int) int {
	sum := 0
	for _, i := range a {
		sum += i
	}
	return sum
}

// --- //

func PowerMod(base, exp, mod int) int {
	result, b, e := 1, base%mod, exp
	for ; e > 0; e, b = e>>1, b*b%mod {
		if e&1 != 0 {
			result = (result * b) % mod
		}
	}
	if result < 0 {
		return result + mod
	}
	return result
}

File: src/main/rust/euler/algorithm/long.rs
// COPYRIGHT (C) 2017 barreiro. All Rights Reserved.
// Rust solvers for Project Euler problems

use algorithm::digits::Digit;
use algorithm::root::{int_sqrt, is_square};

pub trait IncrementAndGet {
    #[allow(clippy::return_self_not_must_use)]
    fn increment_and_get(&mut self) -> Self;
}

impl IncrementAndGet for i32 {
    fn increment_and_get(&mut self) -> Self {
        *self += 1;
        *self
    }
}
impl IncrementAndGet for Digit {
    fn increment_and_get(&mut self) -> Self {
        *self += 1;
        *self
    }
}

impl IncrementAndGet for u64 {
    fn increment_and_get(&mut self) -> Self {
        *self += 1;
        *self
    }
}

impl IncrementAndGet for usize {
    fn increment_and_get(&mut self) -> Self {
        *self += 1;
        *self
    }
}

pub trait GetAndIncrement {
    #[allow(clippy::return_self_not_must_use)]
    fn get_and_increment(&mut self) -> Self;
}

impl GetAndIncrement for i32 {
    fn get_and_increment(&mut self) -> Self {
        let value = *self;
        *self += 1;
        value
    }
}

impl GetAndIncrement for i64 {
    fn get_and_increment(&mut self) -> Self {
        let value = *self;
        *self += 1;
        value
    }
}

impl GetAndIncrement for u64 {
    fn get_and_increment(&mut self) -> Self {
        let value = *self;
        *self += 1;
        value
    }
}

impl GetAndIncrement for usize {
    fn get_and_increment(&mut self) -> Self {
        let value = *self;
        *self += 1;
        value
    }
}

// --- //

/// calculates the Greatest Common Divisor using Stein's algorithm
#[allow(clippy::manual_swap)] // because of const
#[must_use]
pub const fn gcd(mut a: i64, mut b: i64) -> i64 {
    if a == 0 { return b; }
    if b == 0 { return a; }
    let shift = (a | b).trailing_zeros();
    a >>= a.trailing_zeros();
    while b != 0 {
        b >>= b.trailing_zeros();
        if a > b {
            // swap(&mut a, &mut b);
            let tmp = a;
            a = b;
            b = tmp;
        }
        b -= a;
    }
    a << shift
}

/// checks that two values are coprime, i.e. don't have common factors except for 1
#[must_use]
pub const fn are_coprime(a: i64, b: i64) -> bool {
    gcd(a, b) == 1
}

/// calculates the Least Common Multiple
#[must_use]
pub const fn lcm(a: i64, b: i64) -> i64 {
    a / gcd(a, b) * b
}

/// calculates quotient and remainder in one go
#[must_use]
pub const fn div_rem(a: i64, b: i64) -> (i64, i64) {
    (a / b, a % b)
}

/// calculates a quotient that leaves negative remainder
#[must_use]
pub const fn div_ceil(a: i64, b: i64) -> i64 {
    let (q, r) = div_rem(a, b);
    if r == 0 { q } else { q + 1 }
}

/// the sum of all the numbers up to value
#[must_use]
pub const fn arithmetic_sum(value: i64) -> i64 {
    (value * (value + 1)) / 2
}

/// the sum of all the numbers up to value
#[must_use]
pub const fn arithmetic_sum_u64(value: u64) -> u64 {
    (value * (value + 1)) / 2
}

/// simple method to calculate the factorial of small values. No checks are performed. Use with caution.
#[must_use]
pub fn factorial(value: u64) -> u64 {
    (2..=value).product()
}

// --- //

#[must_use]
pub const fn triangle(value: i64) -> i64 {
    arithmetic_sum(value)
}

#[must_use]
pub const fn is_triangle(&value: &i64) -> bool {
    is_square(value * 8 + 1)
}

#[must_use]
pub const fn pentagonal(value: i64) -> i64 {
    (value * (3 * value - 1)) / 2
}

#[must_use]
pub const fn is_pentagonal(&value: &i64) -> bool {
    value == pentagonal(int_sqrt(((value + 1) * 2) / 3))
}

#[must_use]
pub const fn hexagonal(value: i64) -> i64 {
    value * (value * 2 - 1)
}

#[must_use]
pub const fn is_hexagonal(&value: &i64) -> bool {
    value == hexagonal(int_sqrt((value + 1) / 2))
}

#[must_use]
pub const fn heptagonal(value: i64) -> i64 {
    value * (5 * value - 3) / 2
}

#[must_use]
pub const fn octagonal(value: i64) -> i64 {
    value * (3 * value - 2)
}

// --- //

/// efficiently computes `(base ^ exp) mod modulo`
#[must_use]
pub const fn pow_mod(base: u64, exp: u64, modulo: u64) -> u64 {
    let (mut result, mut b, mut e) = (1, base % modulo, exp);
    while e > 0 {
        if e & 1 != 0 {
            result = mul_mod(result, b, modulo);
        }
        e >>= 1;
        b = mul_mod(b, b, modulo);
    }
    result
}

/// computes `(a * b) mod modulo` without overflowing the multiplication
#[must_use]
pub const fn mul_mod(a: u64, mut b: u64, modulo: u64) -> u64 {
    if let Some(result) = a.checked_mul(b) {
        result % modulo
    } else {
        let (mut result, mut j) = (0, a % modulo);
        while b != 0 {
            if b & 1 != 0 {
                result = (result + j) % modulo;
            }
            j = (j * 2) % modulo;
            b >>= 1;
        }
        result % modulo
    }
}
